File: experiments/runs/2026-09-20_07-33-00_claim-office-java-example-mapping_exact-ptdd-v1-cc_opus-5-no-thinking/src/main/java/Reimbursement.java
/**
 * The MHPCO's reimbursement clauses for a damaged item. A deductible of 100 G
 * applies per damage event.
 *
 * <p>Damage to items made of dragon material is fully reimbursed, which is
 * also the MHPCO's default; where both clauses apply, the reduced
 * reimbursement for highly enchanted items wins.
 */
final class Reimbursement {

    private static final int DEDUCTIBLE = 100;
    private static final double REDUCED_REIMBURSEMENT = 0.50;
    private static final int REDUCED_REIMBURSEMENT_LEVEL = 8;

    private Reimbursement() {
    }

    static double forDamage(Item item, int damageAmount) {
        double reimbursement = damageAmount;
        if (isHighlyEnchanted(item)) {
            reimbursement *= REDUCED_REIMBURSEMENT;
        }
        return Math.max(0, reimbursement - DEDUCTIBLE);
    }

    private static boolean isHighlyEnchanted(Item item) {
        return item.enchantment() >= REDUCED_REIMBURSEMENT_LEVEL;
    }
}

File: experiments/runs/2026-09-23_16-10-32_claim-office-example-mapping_exact-ptdd-v1.1-refactor-subagent-cc_opus-5-5-no-thinking/src/reimbursement.ts
import type { InsuredItem } from "./insured-item";

// The MHPCO reimbursement clauses: what a single damage event to an insured item pays out.
const DEDUCTIBLE = 100;
const HIGH_ENCHANTMENT_CLAUSE_LEVEL = 8;
const HIGH_ENCHANTMENT_REIMBURSEMENT_SHARE = 0.5;
const FULL_REIMBURSEMENT_SHARE = 1;

// Claim clause threshold; independent of the premium's high-enchantment surcharge threshold.
function fallsUnderHighEnchantmentClause(damagedItem: InsuredItem): boolean {
  return (damagedItem.enchantment ?? 0) >= HIGH_ENCHANTMENT_CLAUSE_LEVEL;
}

// Which reimbursement clause applies to the damaged item: the share of the damage amount MHPCO covers.
// The high-enchantment clause wins over the dragon-material clause; dragon material is fully reimbursed,
// which coincides with the standard full reimbursement.
function reimbursementShare(damagedItem: InsuredItem): number {
  return fallsUnderHighEnchantmentClause(damagedItem) ? HIGH_ENCHANTMENT_REIMBURSEMENT_SHARE : FULL_REIMBURSEMENT_SHARE;
}

// The deductible is taken per damage event; a damage below the deductible reimburses nothing, never a negative amount.
function afterDeductible(coveredAmount: number): number {
  return Math.max(0, coveredAmount - DEDUCTIBLE);
}

// What MHPCO reimburses for a single damaged item: the covered share first, then the deductible.
export function reimbursement(damageAmount: number, damagedItem: InsuredItem): number {
  return afterDeductible(damageAmount * reimbursementShare(damagedItem));
}
